File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let fast = head
    let slow = head
    for(let i=0; i<n; i++){
        fast = fast.next
    }
    if(!fast){
        return head.next
    }
    while(fast.next){
        fast = fast.next
        slow = slow.next
    }
    slow.next = slow.next.next
    return head
};

// create n gap between fast and slow
// if fast is null which means head node needs to be removed, return head.next
// move the pointers until the end // while fast.next move fast and slow to next
// slow is prev node of target
// update slow.next with slow.next.next
// return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        end = head
        while n:
            end = end.next
            n -= 1        

        start = dummy
        while end:
            start = start.next
            end = end.next
        
        start.next = start.next.next
        return dummy.next

        